lib/evaluate/fold.cc

```diff
@@ -156,35 +162,298 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
+        return Expr<T>{funcRef.arguments()[0]->value->GetType()->kind};
       } else {
         common::die("kind() result not integral");
       }
     } else if (name == "len") {
       if constexpr (std::is_same_v<T, SubscriptInteger>) {
-        if (auto *charExpr{UnwrapExpr<Expr<SomeCharacter>>(*args[0]->value)}) {
+        if (auto *charExpr{UnwrapExpr<Expr<SomeCharacter>>(
+                *funcRef.arguments()[0]->value)}) {
           return std::visit([](auto &kx) { return kx.LEN(); }, charExpr->u);
         }
       } else {
         common::die("len() result not SubscriptInteger");
       }
-    } else {
-      // TODO: many more intrinsic functions
+    } else if (name == "iand") {
+      if (auto *x{std::get_if<BOZLiteralConstant>(
+              &funcRef.arguments()[0]->value->u)}) {
+        *funcRef.arguments()[0]->value =
+            Fold(context, ConvertToType<T>(std::move(*x)));
+      }
+      if (auto *x{std::get_if<BOZLiteralConstant>(
+              &funcRef.arguments()[1]->value->u)}) {
+        *funcRef.arguments()[1]->value =
+            Fold(context, ConvertToType<T>(std::move(*x)));
+      }
+      return FoldElementalIntrinsic<T, T, T>(
+          context, std::move(funcRef), ScalarFunc<T, T, T>(&Scalar<T>::IAND));
+    } else if (name == "int") {
+      return std::visit(
+          [&](auto &&x) -> Expr<T> {
+            using From = std::decay_t<decltype(x)>;
+            if constexpr (std::is_same_v<From, BOZLiteralConstant> ||
+                std::is_same_v<From, Expr<SomeReal>> ||
+                std::is_same_v<From, Expr<SomeInteger>> ||
+                std::is_same_v<From, Expr<SomeComplex>>) {
+              return Fold(context, ConvertToType<T>(std::move(x)));
+            } else {
+              common::die("int() argument type not valid");
+              return Expr<T>{std::move(funcRef)};  // unreachable
+            }
+          },
+          std::move(funcRef.arguments()[0]->value->u));
+    }
+    // TODO: many more intrinsic functions
+  }
+  return Expr<T>{std::move(funcRef)};
+}
+
+template<int KIND>
+Expr<Type<TypeCategory::Real, KIND>> FoldOperation(FoldingContext &context,
+    FunctionRef<Type<TypeCategory::Real, KIND>> &&funcRef) {
+  using namespace intrinsicHelper;
+  using T = Type<TypeCategory::Real, KIND>;
+  for (std::optional<ActualArgument> &arg : funcRef.arguments()) {
+    if (arg.has_value()) {
+      *arg->value = FoldOperation(context, std::move(*arg->value));
+    }
+  }
+  if (auto *intrinsic{std::get_if<SpecificIntrinsic>(&funcRef.proc().u)}) {
+    const std::string name{intrinsic->name};
+    if (name == "acos" || name == "acosh") {
+      if (auto callable{
+              context.hostRte().GetHostProcedureWrapper<Scalar, T, T>(name)}) {
+        return FoldElementalIntrinsic<T, T>(
+            context, std::move(funcRef), *callable);
+      } else {
+        context.messages().Say(
+            "%s(real(kind=%d)) cannot be folded on host"_en_US, name.c_str(),
```

Comment:
Is this a user error or internal error? What does the user do when they get this error?

Reply:
The user code is correct, this was check when resolving intrinsics.  It is reporting to the user an F18 limitation. I think it is hard to guarantee that it will be possible to fold all intrinsics with all possible types on all host architectures.
Pgfortran gives an error in this case (e.g. PGF90-S-0155-Intrinsic not supported in initialization: acosh)
Depending on whether the value is really needed in the front-end or not (e.g. in a kind argument), the compiler may still be able to compile the program or not. That is why it is not yet a hard error.

I am not sure what a user can do. Probably try to adapt his code (I see very little use cases of using an acos with 128bits floating point to get a kind type argument) or add to F18 a host runtime function that can fold this on his host if it is really needed.
Would you rephrase the message to make it clear it is a compiler limitation?

Reply:
If it's a case where the language requires that it be folded to a value, I think it should be a fatal error (like PGI). If we can still compile without folding I'm not sure we need a diagnostic at all. i don't know if you can distinguish those cases.